### PR TITLE
ci(snapshot): skip deploy if no snapshot version or release version

### PR DESCRIPTION
**Pull request by v1v**
### What

- Don't run a snapshot deployment if the version is not a snapshot.
- Don't run a release deployment if the version is a snapshot.

### Why

The release process has been changed to accommodate the branch protection, and hence, the main will contain a release version until the PR with the snapshot version is merged. 

Therefore, the snapshot GitHub action will run when it should not!

In addition, let's avoid the corner case where the `pom.xml` file contains a SNAPSHOT version and the `inputs.ref` points to that particular git reference.

### Test

image

### Issues

Caused by https://github.com/elastic/ecs-logging-java/pull/226

### CODE REVIEW

File: .github/workflows/snapshot.yml [hunk 1]
```diff
@@ -13,10 +13,27 @@ on:
         type: boolean
 
 jobs:
+  validate:
```
[the remark this answers is not in the document]
Comment (v1v) [reply .github/workflows/snapshot.yml:16]: I was initially inclined but since I use the `outputs.is-snapshot` (see `L39`), I somehow thought it was a bit redundant.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -51,6 +51,15 @@ jobs:
             echo "Tag should match pom.xml project.version"
             exit 1
           fi
+      - uses: ./.github/workflows/maven-goal
+        with:
+          command: './mvnw -q help:evaluate -Dexpression=project.version -DforceStdout'
+      - name: Validate version is a release version
+        run: |
+          if [[ "$(./mvnw -q help:evaluate -Dexpression=project.version -DforceStdout)" =~ "-SNAPSHOT" ]]; then
```
Comment (SylvainJuge) [inline .github/workflows/release.yml:59]: This command outputs the following:
```
Apache Maven 3.8.5 (3599d3414f046de2324203b78ddcf9b5e4388aa0)
Maven home: /home/sylvain/.m2/wrapper/dists/apache-maven-3.8.5-bin/5i5jha092a3i37g0paqnfr15e0/apache-maven-3.8.5
Java version: 17.0.8, vendor: Eclipse Adoptium, runtime: /home/sylvain/.sdkman/candidates/java/17.0.8-tem
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.5.0-17-generic", arch: "amd64", family: "unix"
1.47.1-SNAPSHOT%
```

Thus only the last line is relevant here, maybe it's worth adding an `| tail -1` to only get the last line to ensure only the version is checked here, even if it's a bit unlikely that the `SNAPSHOT` string could end up here unless a snapshot maven version is used.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -51,6 +51,15 @@ jobs:
             echo "Tag should match pom.xml project.version"
             exit 1
           fi
+      - uses: ./.github/workflows/maven-goal
```
Comment (JonasKunz) [inline .github/workflows/release.yml:54]: Do we even need this here? Don't we check for exact equality with the provided `end.RELEASE_VERSION` above?

Someone would need to actually provide `1.6.0-SNAPSHOT` here to accidentally release a snapshot?

File: .github/workflows/release.yml (hunk 2 above)
Comment (v1v) [reply .github/workflows/release.yml:59]: I cannot get the same outcome locally:

```bash
$ git --no-pager log -n 1
commit b3d7c6d20ce51e900daee5fb95b4a8a035ff3270 (HEAD -> main, upstream/main, origin/main)

$ ./mvnw -q help:evaluate -Dexpression=project.version -DforceStdout 2> /dev/null
1.6.1-SNAPSHOT

$ ./mvnw -q help:evaluate -Dexpression=project.version -DforceStdout 1> /dev/null
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
```

Comment (v1v) [reply .github/workflows/release.yml:59]: can you confirm whether the stdout contains only the version?

File: .github/workflows/release.yml (hunk 3 above)
Comment (v1v) [reply .github/workflows/release.yml:54]: > Do we even need this here? Don't we check for exact equality with the provided end?RELEASE_VERSION above?

You are right!

Although I'm a bit confused about the reason for using the `inputs.version`, IIUC, it should be the `pre-release` workflow, the one bumping the version and the `release` workflow, the one using the `pom.xml` file, to do the release. Hence `inputs.version` is not something we need at all, the `buildkite` pipeline is not using it.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -51,6 +51,15 @@ jobs:
...
+        with:
+          command: './mvnw -q help:evaluate -Dexpression=project.version -DforceStdout'
```
Comment (v1v) [inline .github/workflows/release.yml:56]:
```suggestion
```

File: .github/workflows/release.yml (hunk 2 above)
Comment (SylvainJuge) [reply .github/workflows/release.yml:56]: I forgot that this depends on the local maven config here https://github.com/elastic/apm-pipeline-library/issues/2186 was the similar issue where we had to deal with the same kind of output. do we have a per-repository maven config that is used for maven options in CI ?

File: .github/workflows/snapshot.yml [hunk 5]
```diff
@@ -13,10 +13,30 @@ on:
...
+    runs-on: ubuntu-latest
+    outputs:
+      is-snapshot: ${{ steps.validate.outputs.is-snapshot }}
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ./.github/workflows/maven-goal
+        with:
+          command: './mvnw -q help:evaluate -Dexpression=project.version -DforceStdout'
```
Comment (v1v) [inline .github/workflows/snapshot.yml:24]: unrequired
```suggestion
```

File: .github/workflows/release.yml (hunk 2 above)
Comment (SylvainJuge) [reply .github/workflows/release.yml:56]: I found where it came from: I was testing on another repository where the maven config actually has the `-V` which makes the output explicitly verbose here, that's something we have to remind removing when opening the PR on the APM agent.

https://github.com/elastic/apm-agent-java/blob/main/.mvn/maven.config

File: .github/workflows/release.yml (hunk 3 above)
Comment (v1v) [reply .github/workflows/release.yml]: > Someone would need to actually provide 1.6.0-SNAPSHOT here to accidentally release a snapshot?

Or change the pom.xml version in a different and use a different `inputs.ref`.

I think the validation is still valid and in fact, the checkout for the given ref is needed, I'll do the change now

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -40,6 +40,7 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v4
         with:
+          ref: ${{ inputs.ref }}
```
Comment (v1v) [inline .github/workflows/release.yml:43]: Fail fast if the ref is not correct. Otherwise, it's required to wait for the Buildkite pipeline to run

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -51,6 +52,12 @@ jobs:
             echo "Tag should match pom.xml project.version"
             exit 1
           fi
+      - name: Validate version is a release version
```
Comment (v1v) [inline .github/workflows/release.yml:55]: Otherwise, the `pom.xml` version is a snapshot.

**Review by JonasKunz — APPROVED** [review]

File: .github/workflows/snapshot.yml [hunk 8]
```diff
@@ -13,10 +13,27 @@ on:
...
+    steps:
+      - uses: actions/checkout@v4
+      - name: Validate version is a snapshot version
+        id: validate
+        run: |
+          output=false
+          if [[ "$(./mvnw -q help:evaluate -Dexpression=project.version -DforceStdout)" =~ "-SNAPSHOT" ]]; then
+            echo "This is a snapshot version"
+            output=true
+          fi
```
Comment (amannocci) [inline .github/workflows/snapshot.yml:29]: What is the advantage of not failing over setting an output and skipping?

Comment (v1v) [reply .github/workflows/snapshot.yml:29]: Otherwise, teams will need to investigate whether it failed for that reason rather than something else.

**Review by amannocci — APPROVED** [review]
LGTM
